File: clients/rust/http.rs
// MiniVault HTTP Client for Rust

use reqwest::{Client, StatusCode};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::time::Duration;

#[derive(Debug, Deserialize)]
pub struct Health {
    pub status: String,
    pub uptime_seconds: i64,
    pub cache_items: i64,
    pub cache_size_mb: i64,
    pub storage_size_mb: i64,
    pub goroutines: i32,
    pub memory_mb: i64,
}

pub struct MiniVault {
    base_url: String,
    api_key: Option<String>,
    client: Client,
}

impl MiniVault {
    pub fn new(base_url: String, api_key: Option<String>) -> Self {
        let client = Client::builder()
            .timeout(Duration::from_secs(5))
            .build()
            .unwrap();

        Self {
            base_url: base_url.trim_end_matches('/').to_string(),
            api_key,
            client,
        }
    }

    pub async fn get(&self, key: &str) -> Result<Option<Vec<u8>>, Box<dyn std::error::Error>> {
        let url = format!("{}/{}", self.base_url, key);
        let response = self.client.get(&url).send().await?;

        match response.status() {
            StatusCode::OK => Ok(Some(response.bytes().await?.to_vec())),
            StatusCode::NOT_FOUND => Ok(None),
            status => Err(format!("GET failed: {}", status).into()),
        }
    }

    pub async fn get_json<T: for<'de> Deserialize<'de>>(
        &self,
        key: &str,
    ) -> Result<Option<T>, Box<dyn std::error::Error>> {
        match self.get(key).await? {
            Some(data) => Ok(Some(serde_json::from_slice(&data)?)),
            None => Ok(None),
        }
    }

    pub async fn set(&self, key: &str, data: Vec<u8>) -> Result<(), Box<dyn std::error::Error>> {
        let url = format!("{}/{}", self.base_url, key);
        let mut request = self.client.put(&url).body(data);

        if let Some(api_key) = &self.api_key {
            request = request.header("X-API-Key", api_key);
        }

        let response = request.send().await?;

        if !response.status().is_success() {
            return Err(format!("SET failed: {}", response.status()).into());
        }

        Ok(())
    }

    pub async fn set_json<T: Serialize>(
        &self,
        key: &str,
        value: &T,
    ) -> Result<(), Box<dyn std::error::Error>> {
        let data = serde_json::to_vec(value)?;
        self.set(key, data).await
    }

    pub async fn delete(&self, key: &str) -> Result<(), Box<dyn std::error::Error>> {
        let url = format!("{}/{}", self.base_url, key);
        let mut request = self.client.delete(&url);

        if let Some(api_key) = &self.api_key {
            request = request.header("X-API-Key", api_key);
        }

        let response = request.send().await?;

        if !response.status().is_success() {
            return Err(format!("DELETE failed: {}", response.status()).into());
        }

        Ok(())
    }

    pub async fn exists(&self, key: &str) -> Result<bool, Box<dyn std::error::Error>> {
        Ok(self.get(key).await?.is_some())
    }

    pub async fn health(&self) -> Result<Health, Box<dyn std::error::Error>> {
        let url = format!("{}/health", self.base_url);
        let response = self.client.get(&url).send().await?;

        if !response.status().is_success() {
            return Err(format!("Health check failed: {}", response.status()).into());
        }

        Ok(response.json().await?)
    }

    pub async fn mget(&self, keys: &[&str]) -> Result<HashMap<String, Vec<u8>>, Box<dyn std::error::Error>> {
        let mut results = HashMap::new();
        let futures: Vec<_> = keys.iter().map(|key| self.get(key)).collect();

        for (i, result) in futures::future::join_all(futures).await.into_iter().enumerate() {
            if let Ok(Some(data)) = result {
                results.insert(keys[i].to_string(), data);
            }
        }

        Ok(results)
    }

    pub async fn mset(&self, entries: &HashMap<&str, Vec<u8>>) -> Result<(), Box<dyn std::error::Error>> {
        let futures: Vec<_> = entries
            .iter()
            .map(|(key, value)| self.set(key, value.clone()))
            .collect();

        for result in futures::future::join_all(futures).await {
            result?;
        }

        Ok(())
    }
}

File: examples/typescript/http.ts
export interface MiniVaultHealth {
  status: string;
  uptime_seconds: number;
  cache_items: number;
  cache_size_mb: number;
  storage_size_mb: number;
  goroutines: number;
  memory_mb: number;
}

export interface MiniVaultOptions {
  baseUrl: string;
  apiKey?: string;
  timeout?: number;
}

export class MiniVault {
  private baseUrl: string;
  private apiKey?: string;
  private timeout: number;

  constructor(options: MiniVaultOptions | string, apiKey?: string) {
    if (typeof options === 'string') {
      this.baseUrl = options;
      this.apiKey = apiKey;
      this.timeout = 5000;
    } else {
      this.baseUrl = options.baseUrl;
      this.apiKey = options.apiKey;
      this.timeout = options.timeout || 5000;
    }
  }

  async get<T = any>(key: string): Promise<T | null> {
    try {
      const url = `${this.baseUrl}/${key}`;
      const controller = new AbortController();
      const timeoutId = setTimeout(() => controller.abort(), this.timeout);

      const headers: HeadersInit = {};
      if (this.apiKey) {
        headers['Authorization'] = `Bearer ${this.apiKey}`;
      }

      const response = await fetch(url, {
        method: 'GET',
        headers,
        signal: controller.signal,
      });

      clearTimeout(timeoutId);

      if (response.status === 404) {
        return null;
      }

      if (!response.ok) {
        return null;
      }

      const result = await response.json();
      if (result.success && result.data !== undefined) {
        return result.data as T;
      }

      return null;
    } catch (error) {
      return null;
    }
  }

  async set(key: string, value: any): Promise<boolean> {
    try {
      const url = `${this.baseUrl}/${key}`;
      const controller = new AbortController();
      const timeoutId = setTimeout(() => controller.abort(), this.timeout);

      const headers: HeadersInit = {
        'Content-Type': 'application/json',
      };

      if (this.apiKey) {
        headers['Authorization'] = `Bearer ${this.apiKey}`;
      }

      const response = await fetch(url, {
        method: 'PUT',
        headers,
        body: JSON.stringify({ value }),
        signal: controller.signal,
      });

      clearTimeout(timeoutId);

      if (!response.ok) {
        return false;
      }

      const result = await response.json();
      return result.success === true;
    } catch (error) {
      return false;
    }
  }

  async delete(key: string): Promise<boolean> {
    try {
      const url = `${this.baseUrl}/${key}`;
      const controller = new AbortController();
      const timeoutId = setTimeout(() => controller.abort(), this.timeout);

      const headers: HeadersInit = {};
      if (this.apiKey) {
        headers['Authorization'] = `Bearer ${this.apiKey}`;
      }

      const response = await fetch(url, {
        method: 'DELETE',
        headers,
        signal: controller.signal,
      });

      clearTimeout(timeoutId);

      const result = await response.json();
      return result.success === true;
    } catch (error) {
      return false;
    }
  }

  async mget<T = any>(keys: string[]): Promise<Array<T | null>> {
    const promises = keys.map(key => this.get<T>(key));
    return Promise.all(promises);
  }

  async mset(entries: Array<{ key: string; value: any }>): Promise<boolean[]> {
    const promises = entries.map(({ key, value }) => this.set(key, value));
    return Promise.all(promises);
  }

  async exists(key: string): Promise<boolean> {
    const data = await this.get(key);
    return data !== null;
  }

  async health(): Promise<MiniVaultHealth | null> {
    try {
      const url = `${this.baseUrl}/health`;
      const controller = new AbortController();
      const timeoutId = setTimeout(() => controller.abort(), this.timeout);

      const response = await fetch(url, { signal: controller.signal });
      clearTimeout(timeoutId);

      if (!response.ok) {
        return null;
      }

      return await response.json() as MiniVaultHealth;
    } catch (error) {
      return null;
    }
  }

  isAvailable(): boolean {
    return !!this.baseUrl;
  }
}

if (typeof module !== 'undefined' && module.exports) {
  module.exports = { MiniVault };
}
